### ISSUE TITLE
Incorrect diagnostic for multiple applicable items some of which are unstable

### ISSUE THREAD

**Issue by djc** [rust-lang/rust#81883]
I tried to compile this code written for Rust 1.47+ with Rust 1.46:

```rust
fn foo(range: Range<usize>) -> bool {
    range.is_empty()
}
```

And I got this error (on 1.46):

```rust
error[E0034]: multiple applicable items in scope
   --> src/lib.rs:133:18
    |
133 |         if range.is_empty() {
    |            ------^^^^^^^^--
    |            |     |
    |            |     multiple `is_empty` found
    |            help: disambiguate the associated function for candidate #2: `std::iter::ExactSizeIterator::is_empty(&range)`
    |
    = note: candidate #1 is defined in an impl for the type `std::ops::Range<Idx>`
    = note: candidate #2 is defined in an impl of the trait `std::iter::ExactSizeIterator` for the type `std::ops::Range<usize>`
```

However, when I tried to use the inherent impl with `Range::is_empty(&range)`, it still failed to compile:

```rust
error[E0034]: multiple applicable items in scope
   --> src/lib.rs:133:19
    |
133 |         if Range::is_empty(range) {
    |            -------^^^^^^^^
    |            |      |
    |            |      multiple `is_empty` found
    |            help: disambiguate the associated function for candidate #2: `std::iter::ExactSizeIterator::is_empty`
    |
    = note: candidate #1 is defined in an impl for the type `std::ops::Range<Idx>`
note: candidate #2 is defined in the trait `std::iter::ExactSizeIterator`
```

As it turns out, `Range::is_empty()` was only stabilized in 1.47.0, so it is not actually available when compiling with stable 1.46.0. So `is_empty()` is arguably not ambiguous here (`ExactSizeIterator` is in the prelude), and the error message is pretty confusing. I think we can do better here!

### Meta

`rustc --version --verbose`:
```
enrai instant-segment $ rustc -vV
rustc 1.46.0
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.46.0
LLVM version: 10.0
```

**Comment by Mark-Simulacrum** [comment]
I think we do better at least in some cases - possibly added since 1.46 though. I think it would be good to have a (multi-crate, unfortunately) reproduction that doesn't depend on std to continue having something unstable.

**Comment by djc** [comment]
How would I reproduce the lack of stability without using std?

**Comment by Mark-Simulacrum** [comment]
`#![feature(staged_api)]` should work to enable any crate to allow declaring things unstable/stable.

**Comment by djc** [comment]
I looked for some documentation on `feature(staged_api)` but the unstable book has an empty entry and some googling didn't turn up anything either. Can someone provide some guidance on how to use that feature to test this?

**Comment by jruderman** [comment]
Same as #63573?

**Comment by djc** [comment]
`staged_api` is documented at https://rustc-dev-guide.rust-lang.org/stability.html#staged_api ([Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/How.20to.20use.20staged_api)). I've been unable to prioritize this for the past 9 months, so if someone else can give it a shot that would be great.
